### PR TITLE
[POC] Shortcut editor (for command mode)

### PR DESCRIPTION
**Pull request by Carreau**
Proof of concept.

1) Ask Webpack to compile es6, because now that we have webpack, why not ? 
2) install react/react-dom.
3) Learn React, and write a module that allows you to edit keyboard shortcuts. 
4) enable eslint for es6.

Todo:
  - Actually persist the new shortcuts
  - actually persist the shortcut we unbind. 
  - design.
  - figure out how to destroy the react components. 
  - do not fail silently if : 
    - shortcut already attached to another action. 
    - Shortcut invalid.
  - decide on JSX.
  - name things correctly.

PIng @rgbkrk and @jdfreder because I have no clue what I am doing. 

---

Personal note, es6 is much less WTF than es5 , but still WTF. 
JSX: replace `<` by `?>` and `/>` by `<?` and you get PHP.

### CODE REVIEW

**Comment by rgbkrk** [comment]
Wow, awesome. I look forward to checking this out.

> JSX: replace < by ?> and /> by <? and you get PHP.

Now you know their secret to success and adoption.

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1 @@
+{ "presets": ["es2015"] }
```
Comment (rgbkrk) [inline .babelrc]: You'll probably want `react` in your presets here, especially if you use JSX. I'm fine either way and won't bikeshed on the [myriad](https://github.com/substack/hyperx) [alternatives](https://github.com/mlmorg/react-hyperscript).

Comment (Carreau) [inline .babelrc]: I don't use JSX for now. I sticked to doing things manually.

Comment (rgbkrk) [inline .babelrc]: Ok, cool!

**Comment by jdfreder** [comment]
Dude awesome!

File: notebook/static/notebook/js/shortcuts_editor.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      handleShrtChange: function (element){
+        this.setState({shrt:element.target.value});
+      },
+      render: function(){
+        const that = this;
+        return React.createElement('div',{style:{borderBottom: '1px solid gray', height: '30px'}},
+                this.props.shortcut? 
+                    React.createElement('i', {className: "pull-right fa fa-times", alt: 'remove title'+this.props.shortcut}):
+                    React.createElement('i', {className: "pull-right fa fa-plus", alt: 'add-keyboard-shortcut', onClick:()=>{
+                        that.props.onAddBindings(that.state.shrt, that.props.ckey);
```
Comment (jdfreder) [inline notebook/static/notebook/js/shortcuts_editor.js]: sweet arrow func

**Comment by jdfreder** [comment]
I'm very happy to see with the webpack stuff the transition to es6 is straight forward :+1:  :+1:  :+1:

**Comment by Carreau** [comment]
> I'm very happy to see with the webpack stuff the transition to es6 is straight forward

The `--watch` is a bit slow and re-parse most of what is in `components`, do you know if there is a way to speed that up ? Or deactivate the optimisations pass on `--watch` ?

File: notebook/static/notebook/js/shortcuts_editor.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                        onClick:()=>{
+                            console.log('will unbind ::', this.props.shortcuts, this.props);
+                            that.props.unbind(this.props.shortcut);
+                        }}):
+                    React.createElement('i', {className: "pull-right fa fa-plus", alt: 'add-keyboard-shortcut',
+                        onClick:()=>{
+                            av?that.props.onAddBindings(that.state.shrt, that.props.ckey):undefined;
+                        }
+                    }),
+                this.props.shortcut? undefined :
```
Comment (rgbkrk) [inline notebook/static/notebook/js/shortcuts_editor.js]: When you don't want an element to show, return `null`.

**Comment by Carreau** [comment]
Closing in favor of #1347 that does not have one of the main concern that goes with the PR.

**Comment by rgbkrk** [comment]
The main concern coming from the dark thread about use of React in Jupyter?

**Comment by Carreau** [comment]
> The main concern coming from the dark thread about use of React in Jupyter?

Yep.

**Pull request closed without merging**